File: backend/handlers/listings.go
package handlers

import (
	"net/http"
	"strconv"
	"strings"
	"time"

	"dumptrade/models"

	"github.com/gin-gonic/gin"
)

type ListingStore interface {
	GetAll(category, status, search, city string, limit, offset int) ([]models.Listing, error)
	GetByID(id int) (*models.Listing, error)
	Create(listing *models.Listing) error
	UpdateStatus(id int, status string) error
	Claim(listingID int, claimantID int) error
	MarkCollected(listingID int, actorID int) error
}

type ListingHandler struct {
	Store ListingStore
}

func NewListingHandler(store ListingStore) *ListingHandler {
	return &ListingHandler{Store: store}
}

func queryInt(c *gin.Context, key string, fallback, max int) int {
	value, err := strconv.Atoi(c.DefaultQuery(key, strconv.Itoa(fallback)))
	if err != nil || value < 0 {
		return fallback
	}
	if max > 0 && value > max {
		return max
	}
	return value
}

func (h *ListingHandler) GetListings(c *gin.Context) {
	category := c.Query("category")
	status := c.Query("status")
	search := strings.TrimSpace(c.Query("search"))
	city := strings.TrimSpace(c.Query("city"))
	if status == "all" {
		status = ""
	}

	limit := queryInt(c, "limit", 24, 100)
	if limit == 0 {
		limit = 24
	}
	offset := queryInt(c, "offset", 0, 0)

	listings, err := h.Store.GetAll(category, status, search, city, limit, offset)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to fetch listings: " + err.Error()})
		return
	}
	c.JSON(http.StatusOK, listings)
}

func (h *ListingHandler) GetListing(c *gin.Context) {
	id, err := strconv.Atoi(c.Param("id"))
	if err != nil || id <= 0 {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid listing ID"})
		return
	}

	listing, err := h.Store.GetByID(id)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to fetch listing: " + err.Error()})
		return
	}
	if listing == nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "Listing not found"})
		return
	}

	c.JSON(http.StatusOK, listing)
}

type CreateListingInput struct {
	Title       string `json:"title" binding:"required"`
	Category    string `json:"category" binding:"required"`
	Description string `json:"description"`
	PhotoURL    string `json:"photo_url"`
	QtyLabel    string `json:"qty_label" binding:"required"`
	QtyNum      int    `json:"qty_num"`
	Condition   string `json:"condition"`
	Location    string `json:"location" binding:"required"`
}

func (h *ListingHandler) CreateListing(c *gin.Context) {
	var input CreateListingInput
	if err := c.ShouldBindJSON(&input); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	input.Title = strings.TrimSpace(input.Title)
	input.QtyLabel = strings.TrimSpace(input.QtyLabel)
	input.Location = strings.TrimSpace(input.Location)
	if input.Title == "" || input.QtyLabel == "" || input.Location == "" {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Title, quantity label, and location are required"})
		return
	}

	validCategories := map[string]bool{
		"furniture": true, "ewaste": true, "textiles": true, "construction": true,
		"organic": true, "plastic": true, "industrial": true, "other": true,
	}
	if !validCategories[input.Category] {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid category"})
		return
	}

	userID, ok := c.Get("userID")
	if !ok {
		c.JSON(http.StatusUnauthorized, gin.H{"error": "User authentication required"})
		return
	}
	userIDInt, ok := userID.(int)
	if !ok || userIDInt <= 0 {
		c.JSON(http.StatusUnauthorized, gin.H{"error": "Invalid authenticated user ID"})
		return
	}

	qtyNum := input.QtyNum
	if qtyNum <= 0 {
		qtyNum = 1
	}

	listing := models.Listing{
		UserID: userIDInt, Title: input.Title, Category: input.Category,
		Description: strings.TrimSpace(input.Description), PhotoUrl: input.PhotoURL,
		QtyLabel: input.QtyLabel, QtyNum: qtyNum, Condition: strings.TrimSpace(input.Condition),
		Location: input.Location, Status: "available", CreatedAt: time.Now(),
	}
	if err := h.Store.Create(&listing); err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to create listing: " + err.Error()})
		return
	}

	c.JSON(http.StatusCreated, listing)
}

func (h *ListingHandler) ClaimListing(c *gin.Context) {
	id, err := strconv.Atoi(c.Param("id"))
	if err != nil || id <= 0 {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid listing ID"})
		return
	}
	userID, ok := c.Get("userID")
	claimantID, validUser := userID.(int)
	if !ok || !validUser || claimantID <= 0 {
		c.JSON(http.StatusUnauthorized, gin.H{"error": "User authentication required"})
		return
	}
	if err := h.Store.Claim(id, claimantID); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}
	c.JSON(http.StatusOK, gin.H{"message": "Listing claimed successfully"})
}

func (h *ListingHandler) MarkCollected(c *gin.Context) {
	id, err := strconv.Atoi(c.Param("id"))
	if err != nil || id <= 0 {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid listing ID"})
		return
	}
	userID, ok := c.Get("userID")
	actorID, validUser := userID.(int)
	if !ok || !validUser || actorID <= 0 {
		c.JSON(http.StatusUnauthorized, gin.H{"error": "User authentication required"})
		return
	}
	if err := h.Store.MarkCollected(id, actorID); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}
	c.JSON(http.StatusOK, gin.H{"message": "Listing marked as collected successfully"})
}

File: api/_lib/routes/listings.js
'use strict';

/**
 * listings.js — "dump material" listings.
 *
 *   GET    /api/listings            ?category & ?status ('all' = no filter) & ?search
 *   GET    /api/listings/:id
 *   POST   /api/listings            (auth)
 *   POST   /api/listings/:id/claim  (auth)  available -> claimed
 *   POST   /api/listings/:id/collect(auth)  claimed   -> collected
 */

const express = require('express');
const { sql } = require('../db');
const { authRequired } = require('../auth');
const { LISTING_SELECT, mapListing } = require('../mappers');
const { asyncRoute, parseId, str, toInt, toBool, body, errors } = require('../http');

const router = express.Router();

const CATEGORIES = new Set([
  'furniture', 'ewaste', 'textiles', 'construction',
  'organic', 'plastic', 'industrial', 'other',
]);

async function getListing(id) {
  const { rows } = await sql.query(`${LISTING_SELECT} WHERE l.id = $1`, [id]);
  return rows[0] ? mapListing(rows[0]) : null;
}

router.get(
  '/',
  asyncRoute(async (req, res) => {
    const category = str(req.query.category);
    const statusRaw = str(req.query.status);
    const status = statusRaw === 'all' ? '' : statusRaw;
    const search = str(req.query.search);

    // sql.query() parameterises exactly like the sql`` tag; it is used here
    // because the filters are optional. The ::text casts let Postgres infer the
    // parameter types in the `$n = ''` comparisons.
    const { rows } = await sql.query(
      `${LISTING_SELECT}
       WHERE ($1::text = '' OR l.category::text = $1::text)
         AND ($2::text = '' OR l.status::text = $2::text)
         AND ($3::text = '' OR l.title ILIKE '%' || $3::text || '%'
                            OR COALESCE(l.description, '') ILIKE '%' || $3::text || '%'
                            OR COALESCE(l.location, '') ILIKE '%' || $3::text || '%')
       ORDER BY l.created_at DESC, l.id DESC`,
      [category, status, search]
    );

    return res.json(rows.map(mapListing));
  })
);

router.get(
  '/:id',
  asyncRoute(async (req, res) => {
    const id = parseId(req.params.id);
    if (!id) return errors.badRequest(res, 'Invalid listing ID');

    const listing = await getListing(id);
    if (!listing) return errors.notFound(res, 'Listing not found');
    return res.json(listing);
  })
);

router.post(
  '/',
  authRequired,
  asyncRoute(async (req, res) => {
    const input = body(req);
    const title = str(input.title);
    const category = str(input.category) || 'other';
    const qtyLabel = str(input.qty_label);
    const location = str(input.location);

    if (!title) return errors.badRequest(res, 'title is required');
    if (!CATEGORIES.has(category)) {
      return errors.badRequest(
        res,
        'Invalid category. Must be one of: furniture, ewaste, textiles, construction, organic, plastic, industrial, other'
      );
    }
    if (!qtyLabel) return errors.badRequest(res, 'qty_label is required');
    if (!location) return errors.badRequest(res, 'location is required');

    const qtyNum = Math.max(1, toInt(input.qty_num, 1));

    const { rows } = await sql`
      INSERT INTO listings
        (user_id, title, category, description, photo_url, qty_label, qty_num,
         condition, location, status, needs_disposer, disposer_note)
      VALUES
        (${req.userID}, ${title}, ${category}, ${str(input.description)}, ${str(input.photo_url)},
         ${qtyLabel}, ${qtyNum}, ${str(input.condition)}, ${location}, 'available',
         ${toBool(input.needs_disposer)}, ${str(input.disposer_note)})
      RETURNING id`;

    const listing = await getListing(rows[0].id);
    return res.status(201).json(listing);
  })
);

/** available -> claimed. The conditional UPDATE makes the race atomic. */
router.post(
  '/:id/claim',
  authRequired,
  asyncRoute(async (req, res) => {
    const id = parseId(req.params.id);
    if (!id) return errors.badRequest(res, 'Invalid listing ID');

    const { rows } = await sql`
      UPDATE listings SET status = 'claimed'
      WHERE id = ${id} AND status::text = 'available'
      RETURNING id`;

    if (!rows.length) {
      const current = await sql`SELECT status::text AS status FROM listings WHERE id = ${id}`;
      if (!current.rows.length) return errors.notFound(res, 'Listing not found');
      return errors.conflict(res, `Sorry — this listing is already ${current.rows[0].status}.`);
    }

    await sql`INSERT INTO claims (listing_id, claimant_id) VALUES (${id}, ${req.userID})`;

    return res.json({
      message: 'Listing claimed successfully',
      listing: await getListing(id),
    });
  })
);

/** claimed -> collected. */
router.post(
  '/:id/collect',
  authRequired,
  asyncRoute(async (req, res) => {
    const id = parseId(req.params.id);
    if (!id) return errors.badRequest(res, 'Invalid listing ID');

    const { rows } = await sql`
      UPDATE listings SET status = 'collected'
      WHERE id = ${id} AND status::text = 'claimed'
      RETURNING id`;

    if (!rows.length) {
      const current = await sql`SELECT id FROM listings WHERE id = ${id}`;
      if (!current.rows.length) return errors.notFound(res, 'Listing not found');
      return errors.conflict(res, "This item isn't awaiting collection.");
    }

    await sql`
      UPDATE claims SET collected_at = now()
      WHERE listing_id = ${id} AND collected_at IS NULL`;

    return res.json({
      message: 'Listing marked as collected successfully',
      listing: await getListing(id),
    });
  })
);

module.exports = router;
